### ISSUE TITLE
BUG: workqueue lockup - pool cpus=2 node=0 flags=0x0 nice=-20 stuck for 58s!

### ISSUE THREAD

**Issue by marcin-github** [koverstreet/bcachefs#37]
commit 990250b29634fa59d8689f49a9fd7b5d4c134634 (but it appears also on earlier commits)

```
2019-04-24T11:34:00.514575+02:00 jowisz kernel: [  752.907614] BUG: workqueue lockup - pool cpus=2 node=0 flags=0x0 nice=-20 stuck for 58s!
2019-04-24T11:34:00.514655+02:00 jowisz kernel: [  752.914030] Showing busy workqueues and worker pools:
2019-04-24T11:34:00.514659+02:00 jowisz kernel: [  752.914034] workqueue events_highpri: flags=0x10
2019-04-24T11:34:00.514661+02:00 jowisz kernel: [  752.914036]   pwq 5: cpus=2 node=0 flags=0x0 nice=-20 active=2/256
2019-04-24T11:34:00.514668+02:00 jowisz kernel: [  752.914042]     pending: journal_write_done [bcachefs], bch2_journal_write [bcachefs]
2019-04-24T11:34:00.514677+02:00 jowisz kernel: [  752.914131] workqueue writeback: flags=0x4e
2019-04-24T11:34:00.514680+02:00 jowisz kernel: [  752.914132]   pwq 8: cpus=0-3 flags=0x4 nice=0 active=3/256
2019-04-24T11:34:00.514681+02:00 jowisz kernel: [  752.914135]     in-flight: 11518:wb_workfn, 1705:wb_workfn
2019-04-24T11:34:00.515382+02:00 jowisz kernel: [  752.914188]     pending: wb_workfn
2019-04-24T11:34:00.515387+02:00 jowisz kernel: [  752.914191] workqueue kblockd: flags=0x18
2019-04-24T11:34:00.515388+02:00 jowisz kernel: [  752.914192]   pwq 1: cpus=0 node=0 flags=0x0 nice=-20 active=1/256
2019-04-24T11:34:00.515389+02:00 jowisz kernel: [  752.914195]     pending: blk_mq_timeout_work
2019-04-24T11:34:00.515391+02:00 jowisz kernel: [  752.914241] workqueue bcachefs: flags=0x1c
2019-04-24T11:34:00.515392+02:00 jowisz kernel: [  752.914242]   pwq 5: cpus=2 node=0 flags=0x0 nice=-20 active=1/1
2019-04-24T11:34:00.515393+02:00 jowisz kernel: [  752.914244]     in-flight: 95:bch2_write_index [bcachefs]
2019-04-24T11:34:00.515394+02:00 jowisz kernel: [  752.914258]     delayed: bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs],
bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs]
2019-04-24T11:34:00.515415+02:00 jowisz kernel: [  752.914635] , bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs]

[...]

2019-04-24T11:34:00.515843+02:00 jowisz kernel: [  752.931174] , bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs]
2019-04-24T11:34:00.515847+02:00 jowisz kernel: [  752.931559] , bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs], bch2_write_index [bcachefs]
2019-04-24T11:34:00.515860+02:00 jowisz kernel: [  752.931712] workqueue bcache_journal: flags=0x1c
2019-04-24T11:34:00.515865+02:00 jowisz kernel: [  752.931714]   pwq 5: cpus=2 node=0 flags=0x0 nice=-20 active=1/1
2019-04-24T11:34:00.515867+02:00 jowisz kernel: [  752.931718]     pending: bch2_journal_reclaim_work [bcachefs]
2019-04-24T11:34:00.515877+02:00 jowisz kernel: [  752.931749] workqueue kcryptd/253:5: flags=0x3a
2019-04-24T11:34:00.515881+02:00 jowisz kernel: [  752.931750]   pwq 9: cpus=0-3 flags=0x4 nice=-20 active=2/4
2019-04-24T11:34:00.515883+02:00 jowisz kernel: [  752.931753]     pending: kcryptd_crypt [dm_crypt], kcryptd_crypt [dm_crypt]
2019-04-24T11:34:00.515886+02:00 jowisz kernel: [  752.931827] pool 5: cpus=2 node=0 flags=0x0 nice=-20 hung=58s workers=4 idle: 16426 13865 23
2019-04-24T11:34:00.515888+02:00 jowisz kernel: [  752.931845] pool 8: cpus=0-3 flags=0x4 nice=0 hung=0s workers=4 idle: 367 2988
```

**Comment by marcin-github** [comment]
```
# zgrep -iE "(throt|writeback|wbt)" /proc/config.gz
CONFIG_CGROUP_WRITEBACK=y
CONFIG_BLK_DEV_THROTTLING=y
CONFIG_BLK_DEV_THROTTLING_LOW=y
CONFIG_BLK_WBT=y
CONFIG_BLK_WBT_MQ=y
# CONFIG_ZRAM_WRITEBACK is not set
```

**Comment by koverstreet** [comment]
This should be fixed by f950e2123a.

There's still another bug in that sometimes we're spinning and burning cpu in the btree update path when we shouldn't be - specifically, when we should be blocked waiting for space in the journal.

That spinning was interacting badly with a workqueue implementation detail, which was causing other workqueue items to be on different workqueues to be blocked (which normally should not happen!) - including work items that needed to complete for the btree update to make forward progress, leading to deadlock.

The patch I mentioned makes it so the work items won't get blocked and we won't deadlock. Still have to figure out why we're spinning and burning CPU, but it should now be more of annoyance than anything fatal.

**Comment by marcin-github** [comment]
Indeed, since a couple days I don't have this issue anymore. Thank you.
